persim/images.py

```diff
@@ -210,3 +220,695 @@ def show(self, imgs, ax=None):
+                 weight=None, weight_params=None, kernel=None, kernel_params=None):
+        """
+        Class for transforming persistence diagrams into persistence images
+        """
+        # set defaults
+        if birth_range is None:
+            birth_range = (0.0, 1.0)
+        if pers_range is None:
+            pers_range = (0.0, 1.0)
+        if pixel_size is None:
+            pixel_size = np.min([pers_range[1] - pers_range[0], birth_range[1] - birth_range[0]]) * 0.2
+        if weight is None:
+            weight = persistence
+        if kernel is None:
+            kernel = bvncdf
+        if weight_params is None:
+            weight_params = {'n': 1.0}
+        if kernel_params is None:
+            kernel_params = {'sigma': np.array([[1.0, 0.0], [0.0, 1.0]])}
+       
+        self.weight = weight
+        self.weight_params = weight_params
+        self.kernel = kernel
+        self.kernel_params = kernel_params
+        self._pixel_size = pixel_size
+        self._birth_range = birth_range
+        self._pers_range = pers_range
+        self._width = birth_range[1] - birth_range[0]
+        self._height = pers_range[1] - pers_range[0]
+        self._resolution = (int(self._width / self._pixel_size), int(self._height / self._pixel_size))
+        self._create_mesh()
+
+    @property
+    def width(self):
+        return self._width
+
+    @property
+    def height(self):
+        return self._height
+
+    @property
+    def resolution(self):
+        return self._resolution
+
+    @property
+    def pixel_size(self):
+        return self._pixel_size
+
+    @pixel_size.setter
+    def pixel_size(self, val):
+        self._pixel_size = val
+        self._width = int(np.ceil((self.birth_range[1] - self.birth_range[0]) / self.pixel_size)) * self.pixel_size
+        self._height = int(np.ceil((self.pers_range[1] - self.pers_range[0]) / self.pixel_size)) * self.pixel_size
+        self._resolution = (int(self.width / self.pixel_size), int(self.height / self.pixel_size))
+        self._create_mesh()
+
+    @property
+    def birth_range(self):
+        return self._birth_range
+
+    @birth_range.setter
+    def birth_range(self, val):
+        self._birth_range = val
+        self._width = int(np.ceil((self.birth_range[1] - self.birth_range[0]) / self.pixel_size)) * self._pixel_size
+        self._resolution = (int(self.width / self.pixel_size), int(self.height / self.pixel_size))
+        self._create_mesh()
+
+    @property
+    def pers_range(self):
+        return self._pers_range
+
+    @pers_range.setter
+    def pers_range(self, val):
+        self._pers_range = val
+        self._height = int(np.ceil((self.pers_range[1] - self.pers_range[0]) / self.pixel_size)) * self._pixel_size
+        self._resolution = (int(self.width / self.pixel_size), int(self.height / self.pixel_size))
+        self._create_mesh()
+
+    def __repr__(self):
+        repr_str = 'PersistenceImager object: \n' +\
```

Comment:
Could you please format the repr so that the print results creates a valid constructor similar to how scikit-learn structures theres.. I think it would look more like
```
PersistenceImager(pixel_size={}, ...)
```

Reply:
PersistenceImager().__repr__ has been updated to create a valid constructor.  

A few observations and thoughts: To make use of the functionality already implemented in scikit learn transformers, it looks like one would at least need to include both TransformerMixin and BaseEstimator in the class definition, since for some reason NewClass(TransformerMixin) on its own does not produce the same results with regards to the inherited __repr__ as  NewClass(TransformerMixin, BaseEstimator). For example, see MinMaxScaler here https://github.com/scikit-learn/scikit-learn/blob/42aff4e2e/sklearn/preprocessing/_data.py. 

Since the PersistenceImager() class has attributes which are expected to be callable kernel and weight functions, I overrode any inheritance of __repr__ from TransformerMixin to ensure a proper constructor is made, since the scikit learn method will produce a none readable string (according to the pprint.isreadable()) that references the function's memory address.  As a consequence, this means that one must ensure the callable kernel and weight functions are in the local namespace for the output of PersistenceImager().__repr__ to actually be able to work as a constructor. I suppose this could be avoided in part by making the class allow for a string to be used to specify the kernel and weight functions, with additional methods in PersistenceImager() to handle conversion from string to valid callable, but then there would be conflict and exceptions needed for custom kernels. As it is now, the NameError produced by 
```
>>> pimg=PersistenceImager()
>>> eval(repr(pimg)) 
         NameError: name 'persistence' is not defined
```
is quite transparent in my opinion and is easily fixed:
```
>>> from persim.images_weights import persistence
>>> from persim.images_kernels import gaussian
>>> pimg=PersistenceImager()
>>> eval(repr(pimg)) 
```
So I am inclined to not overly complicate the inputs to the kernel and weight arguments.

Reply:
Sounds like this is a bit of a rabbit hole?  In the past, we've used set it so the arguments are string or callable, with the string keys pointing to the internally defined functions. I think UMAP follows this convention for it's distance functions. 

Whatever is easiest for you at this point.  The reformatting of the repr already accommodated my main nit.

Reply:
On second thought, it was pretty clean to implement this in a similar manner to UMAP. Latest update now allowes for either a valid string or a callable.  I also added a parameter validation method (inspired by UMAP) and tests for these.  Seems pretty easily extensible to other predefined weights or kernels if so desired at some point in the future.